File: core/src/main/java/dev/vibris/core/ShaderReloadFailure.java
package dev.vibris.core;

import dev.vibris.api.ReloadResult;
import dev.vibris.protocol.v1.ArtifactMetadata;
import dev.vibris.protocol.v1.ErrorCode;

import java.io.IOException;

final class ShaderReloadFailure {
    private ShaderReloadFailure() {
    }

    static RuntimeJobExecutor.Failure create(ShaderLogSink logs, CoreJob job, ReloadResult reload) {
        String message = reload.diagnostics().stream()
            .filter(diagnostic -> diagnostic.severity() == ReloadResult.Severity.ERROR)
            .map(ReloadResult.Diagnostic::message)
            .findFirst()
            .orElse("Shader reload failed.");
        try {
            ArtifactMetadata artifact = logs.writeShaderLog(
                job.workspaceId, job.requestId, reload.diagnostics());
            return new RuntimeJobExecutor.Failure(ErrorCode.SHADER_COMPILE_FAILED, message, artifact);
        } catch (IOException exception) {
            return new RuntimeJobExecutor.Failure(ErrorCode.SHADER_COMPILE_FAILED, message);
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/ShaderReloadFailure.kt
package dev.vibris.core

import dev.vibris.api.ReloadResult
import dev.vibris.protocol.v2.ErrorCode
import java.io.IOException

internal object ShaderReloadFailure {
    @JvmStatic
    fun create(
        logs: ShaderLogSink,
        job: CoreJob,
        reload: ReloadResult,
    ): RuntimeJobExecutor.Failure {
        val message = reload.diagnostics
            .firstOrNull { it.severity == ReloadResult.Severity.ERROR }
            ?.message
            ?: "Shader reload failed."
        return try {
            val artifact = logs.writeShaderLog(job.workspaceId, job.requestId, reload.diagnostics)
            RuntimeJobExecutor.Failure(
                ErrorCode.ERROR_CODE_SHADER_COMPILE_FAILED,
                message,
                java.util.List.of(artifact),
                reload.diagnostics,
                null,
                false,
            )
        } catch (_: IOException) {
            RuntimeJobExecutor.Failure(
                ErrorCode.ERROR_CODE_SHADER_COMPILE_FAILED,
                message,
                java.util.List.of(),
                reload.diagnostics,
                null,
                false,
            )
        }
    }
}
